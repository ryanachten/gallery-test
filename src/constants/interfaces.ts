export type Action = {
  type: string;
  payload?: any;
  error?: any;
};

export enum Size {
  ALL = "",
  XS = "XS",
  S = "S",
  M = "M",
  L = "L",
  XL = "XL"
}

export type Product = {
  index: number;
  isSale: boolean;
  isExclusive: boolean;
  price: string;
  productImage: string;
  productName: string;
  size: Size[];
};
