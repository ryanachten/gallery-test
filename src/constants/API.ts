export const API_URL = "https://api.jsonbin.io/b/5cae9a54fb42337645ebcad3";
