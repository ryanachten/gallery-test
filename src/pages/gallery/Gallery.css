.Gallery__header {
  align-items: center;
  background-color: #def1f4;
  display: flex;
  justify-content: space-between;
  margin-bottom: 0.31em;
  padding: 0.31em;
}

.Gallery__title {
  font-size: 2.25em;
}

.Gallery__filter {
  height: fit-content;
}

.Gallery__grid {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
}
