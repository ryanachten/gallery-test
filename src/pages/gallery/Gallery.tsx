import React, { Component, ChangeEvent } from "react";
import { connect } from "react-redux";
import ProductThumbnail from "../../components/gallery/ProductThumbnail";
import { Product, Size } from "../../constants/interfaces";
import { productActions } from "../../reducers/product";
import "./Gallery.css";

type Props = {
  getProducts: () => Promise<void>;
  products: Product[];
};

type State = {
  filteredProducts: Product[];
  selectedSize: Size;
};

export class Gallery extends Component<Props, State> {
  sizes: Size[];
  constructor(props: Props) {
    super(props);
    this.state = {
      filteredProducts: [],
      selectedSize: Size.ALL
    };
    this.getProducts = this.getProducts.bind(this);
    this.onFilterChange = this.onFilterChange.bind(this);
    this.sizes = [Size.XS, Size.S, Size.M, Size.L, Size.XL];
  }

  componentDidMount() {
    this.getProducts();
  }

  async getProducts() {
    await this.props.getProducts();
    this.setState({
      filteredProducts: this.props.products
    });
  }

  onFilterChange(e: ChangeEvent<HTMLSelectElement>) {
    const value: any = e.target.value;
    const products = this.props.products;
    // If the filter is not set, we reset the filtered products
    // Otherwise, we filter the products by whether or not they
    // have the have the selected size available
    const filteredProducts: Product[] =
      value === Size.ALL
        ? products
        : products.filter((product: Product) => product.size.includes(value));

    this.setState({
      filteredProducts,
      selectedSize: value
    });
  }

  render() {
    const { filteredProducts, selectedSize } = this.state;
    return (
      <div className="Gallery">
        <header className="Gallery__header">
          <h1 className="Gallery__title">Women's tops</h1>
          <select
            aria-label="filter"
            className="Gallery__filter"
            onChange={this.onFilterChange}
            value={selectedSize}
          >
            <option value={Size.ALL}>
              {selectedSize === Size.ALL ? "Filter by size" : "Clear filter"}
            </option>
            {this.sizes.map(size => (
              <option key={size} value={size}>
                {size}
              </option>
            ))}
          </select>
        </header>
        <main className="Gallery__grid">
          {filteredProducts.length > 0 &&
            filteredProducts.map((product: Product) => (
              <ProductThumbnail key={product.index} product={product} />
            ))}
        </main>
      </div>
    );
  }
}

const mapStateToProps = ({ product }: any) => ({
  products: product.products
});

const mapDispatchToProps = {
  getProducts: productActions.getProducts
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Gallery);
