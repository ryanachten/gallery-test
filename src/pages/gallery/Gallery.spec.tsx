import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { Gallery } from "./Gallery";
import { Product, Size } from "../../constants/interfaces";

describe("Page: Gallery", () => {
  const initProps = {
    getProducts: jest.fn(),
    products: []
  };

  const productTemplate: Product = {
    index: 0,
    isSale: true,
    isExclusive: false,
    price: "$18.88",
    productImage: "product-1.jpg",
    productName: "Striped shirt",
    size: [Size.XS, Size.S, Size.L, Size.XL]
  };

  it("renders correctly", () => {
    const tree = renderer.create(<Gallery {...initProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should call getProducts on mount", () => {
    const products = [{ ...productTemplate }];
    const getProducts = jest.fn().mockImplementation(() => products);
    shallow(<Gallery {...initProps} getProducts={getProducts} />);
    expect(getProducts).toHaveBeenCalled();
  });

  it("should set filtered products on filter change", () => {
    const smallProduct = { ...productTemplate, size: [Size.S] };
    const largeProduct = { ...productTemplate, size: [Size.L] };
    const products = [smallProduct, largeProduct];
    const wrapper = shallow(<Gallery {...initProps} products={products} />);
    const instance: any = wrapper.instance();

    // Scenario 1: should filter only small products
    instance.onFilterChange({
      target: {
        value: Size.S
      }
    });
    expect(instance.state.filteredProducts).toEqual([smallProduct]);

    // Scenario 2: should filter only large products
    instance.onFilterChange({
      target: {
        value: Size.L
      }
    });
    expect(instance.state.filteredProducts).toEqual([largeProduct]);

    // Scenario 3: should clear filter to view to all products
    instance.onFilterChange({
      target: {
        value: Size.ALL
      }
    });
    expect(instance.state.filteredProducts).toEqual(products);
  });
});
