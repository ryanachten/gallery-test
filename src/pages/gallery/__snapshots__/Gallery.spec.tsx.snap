// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page: Gallery renders correctly 1`] = `
<div
  className="Gallery"
>
  <header
    className="Gallery__header"
  >
    <h1
      className="Gallery__title"
    >
      Women's tops
    </h1>
    <select
      aria-label="filter"
      className="Gallery__filter"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        Filter by size
      </option>
      <option
        value="XS"
      >
        XS
      </option>
      <option
        value="S"
      >
        S
      </option>
      <option
        value="M"
      >
        M
      </option>
      <option
        value="L"
      >
        L
      </option>
      <option
        value="XL"
      >
        XL
      </option>
    </select>
  </header>
  <main
    className="Gallery__grid"
  />
</div>
`;
